src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,810 @@
+  }
+  ```
+
+* Preserved array $values property contains null
+  * **Newtonsoft.Json**: Throw - Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.
+
+  ```json
+  {
+      "$id": "1",
+      "$values": null
+  }
+  ```
+
+* Preserved array $values property contains value
+  * **Newtonsoft.Json**: Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - The JSON value could not be converted to TArray. Path: $.$values
+
+  ```json
+  {
+      "$id": "1",
+      "$values": 1
+  }
+  ```
+
+* Preserved array $values property contains object
+  * **Newtonsoft.Json**: Unexpected token while deserializing object: EndObject. Path ''.
+  * **S.T.Json**: Throw - The property is already part of a preserved array object, cannot be read as a preserved array.
+
+  ```json
+  {
+      "$id": "1",
+      "$values": {}
+  }
+  ```
+
+## JSON Objects if not Collection (Class | Struct | Dictionary) - On Deserialize (and Serialize?)
+
+* `$ref` **Valid** under conditions:
+  * must be the only property in the object.
+
+* `$id` **Valid** under conditions:
+  * must be the first property in the object
+
+* `$values` **Not Valid**
+
+* `$.*` **Valid**
+
+Note: For Dictionary keys on serialize, should we allow serializing keys `$id`, `$ref` and `$values`? if we allow it, then there is a potential round-tripping issue.
+Sample of similar issue with `DictionaryKeyPolicy`:
+```cs
+public static void TestDictionary_Collision() 
+{
+    var root = new Dictionary<string, int>();
+    root["helloWorld"] = 100;
+    root["HelloWorld"] = 200;
+
+    var opts = new JsonSerializerOptions
+    {
+        DictionaryKeyPolicy = JsonNamingPolicy.CamelCase
+    };
+
+    string json = JsonSerializer.Serialize(root, opts);
+    Console.WriteLine(json);
+    /* Output:
+    {"helloWorld":100,"helloWorld":200} */
+
+    // Round tripping issue
+    root = JsonSerializer.Deserialize<Dictionary<string, int>>(json);
+}
+```
+
+
+## JSON Object if Collection - On Deserialize
+
+* `$ref` **Valid** under conditions:
+  * must be the only property in the object.
+
+* `$id` **Valid** under conditions:
+  * must be the first property in the object.
+
+* `$values` **Valid** under conditions:
+  * must be after `$id`
+
+* `$.*` **Not Valid**
+
+
+## Immutable types
+Since these types are created with the help of an internal converter, and they are not parsed until the entire block of JSON finishes; nested reference to these types is impossible to identify, unless you re-scan the resulting object, which is too expensive.
+
+With that said, the deserializer will throw when it reads `$id` on any of these types; but regardless of that, when writing those types, they are going to be preserved as any other collection type (`{ "$id": "1", "$values": [...] }`) since those types can still being parsed into a collection type that it is supported.
```

Comment:
This means the `JsonSerializer` will produce a payload that the deserializer can't process (won't round-trip). Please make a note of that.

```suggestion
With that said, the deserializer will throw when it reads `$id` on any of these types. When serializing (i.e. writing) those types, however, they are going to be preserved as any other collection type (`{ "$id": "1", "$values": [...] }`) since those types can still be parsed into a collection type that is supported.
```